### PR TITLE
Add support for arbitrary single-letter tags in Filter

### PR DESCRIPTION
**Pull request by tyiu**
https://github.com/nostr-protocol/nips/blob/4fa65b81cede9eaf8375d4f92bc80cf298190e0c/01.md#from-client-to-relay-sending-events-and-creating-subscriptions

### CODE REVIEW

File: Sources/NostrSDK/Filter.swift [hunk 1]
```diff
@@ -10,7 +10,7 @@ import Foundation
 /// A structure that describes a filter to subscribe to relays with.
 ///
 /// > Note: [NIP-01 Specification](https://github.com/nostr-protocol/nips/blob/master/01.md#communication-between-clients-and-relays)
-public struct Filter: Codable, Hashable, Equatable {
+public struct Filter: Encodable, Hashable, Equatable {
```
Comment (tyiu) [inline Sources/NostrSDK/Filter.swift:13]: We don't decode Filter anywhere, so I changed it to be only `Encodable`.

File: Sources/NostrSDK/Filter.swift [hunk 2]
```diff
@@ -54,19 +63,55 @@ public struct Filter: Codable, Hashable, Equatable {
     ///   - kinds: a list of a kind numbers
     ///   - events: a list of event ids that are referenced in an "e" tag
     ///   - pubkeys: a list of pubkeys that are referenced in a "p" tag
+    ///   - tags: a list of tag values that are referenced by single English-alphabet letter tag names
     ///   - since: an integer unix timestamp, events must be newer than this to pass
     ///   - until: an integer unix timestamp, events must be older than this to pass
     ///   - limit: maximum number of events to be returned in the initial query
     ///
+    /// If `tags` contains an `e` tag and `events` is also provided, `events` takes precedence.
```
Comment (tyiu) [inline Sources/NostrSDK/Filter.swift:71]: I kept in `events` and `pubkeys` for backwards compatibility, but also developer ergonomics. Let me know if you think this is a bad idea.

**Review by bryanmontz — CHANGES_REQUESTED** [review]

File: Sources/NostrSDK/Filter.swift [hunk 3]
```diff
@@ -36,16 +33,28 @@ public struct Filter: Codable, Hashable, Equatable {
...
+    private struct TagFilterName: CodingKey {
+        var stringValue: String
+
+        init(stringValue: String) {
+            self.stringValue = stringValue
+        }
+
+        var intValue: Int? { nil }
+
+        init?(intValue: Int) {
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Filter.swift:53]: It's declaration is required by the `CodingKey` protocol. The compiler doesn't like it if it's not there. Per the documentation of this initializer, it says

```swift
    /// Creates a new instance from the specified integer.
    ///
    /// If the value passed as `intValue` does not correspond to any instance of
    /// this type, the result is `nil`.
    ///
    /// - parameter intValue: The integer value of the desired key.
    init?(intValue: Int)
```

File: Sources/NostrSDK/Filter.swift [hunk 4]
```diff
@@ -54,19 +63,85 @@ public struct Filter: Codable, Hashable, Equatable {
...
+
+        var tagsContainer = encoder.container(keyedBy: TagFilterName.self)
+        try self.tags?.forEach {
+            try tagsContainer.encode($0.value, forKey: TagFilterName(stringValue: "#\($0.key)"))
+        }
+    }
+}
+
+private extension Character {
+    var isEnglishLetter: Bool {
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Filter.swift:144]: I think "Latin script" might be too broad.

>The Latin script is the basis of the [International Phonetic Alphabet](https://en.wikipedia.org/wiki/International_Phonetic_Alphabet), and the 26 most widespread letters are the letters contained in the [ISO basic Latin alphabet](https://en.wikipedia.org/wiki/ISO_basic_Latin_alphabet), which are the same letters as the [English alphabet](https://en.wikipedia.org/wiki/English_alphabet).

Maybe `isBasicLatinAlphabet` is better.

File: Sources/NostrSDK/Filter.swift [hunk 5]
```diff
@@ -19,13 +19,10 @@ public struct Filter: Codable, Hashable, Equatable {
...
     public let kinds: [Int]?
-    
-    /// a list of event ids that are referenced in an "e" tag
-    public let events: [String]?
-    
-    /// a list of pubkeys that are referenced in a "p" tag
-    public let pubkeys: [String]?
-    
+
+    /// a list of tag values that are referenced by single English-alphabet letter tag names
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Filter.swift]: Changed it to "basic Latin letter" instead.
https://en.wikipedia.org/wiki/ISO_basic_Latin_alphabet

File: Sources/NostrSDK/Filter.swift [hunk 6]
```diff
@@ -54,19 +63,85 @@ public struct Filter: Codable, Hashable, Equatable {
     ///   - kinds: a list of a kind numbers
     ///   - events: a list of event ids that are referenced in an "e" tag
     ///   - pubkeys: a list of pubkeys that are referenced in a "p" tag
+    ///   - tags: a list of tag values that are referenced by single English-alphabet letter tag names
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Filter.swift]: Changed it to "basic Latin letter" instead.
https://en.wikipedia.org/wiki/ISO_basic_Latin_alphabet

File: Sources/NostrSDK/Filter.swift [hunk 7]
```diff
@@ -54,19 +63,85 @@ public struct Filter: Codable, Hashable, Equatable {
...
     ///   - since: an integer unix timestamp, events must be newer than this to pass
     ///   - until: an integer unix timestamp, events must be older than this to pass
     ///   - limit: maximum number of events to be returned in the initial query
     ///
+    /// If `tags` contains an `e` tag and `events` is also provided, `events` takes precedence.
+    /// If `tags` contains a `p` tag and `pubkeys` is also provided, `pubkeys` takes precedence.
+    ///
+    /// Returns `nil` if `tags` contains tag names that are not in the English-alphabet of A-Z or a-z.
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Filter.swift]: Changed it to "basic Latin letter" instead.
https://en.wikipedia.org/wiki/ISO_basic_Latin_alphabet

**Review by bryanmontz — APPROVED** [review]

**Comment by bryanmontz** [comment]
@tyiu Feel free to merge this one after resolving conflicts.

**Pull request merged**